File: Cpp/day_2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>

bool symmetry_checker(const std::string& s) {
    int half = s.length() / 2;
    if (half > 0 && s.substr(0, half) == s.substr(half)) {
        return true;
    }
    return false;
}

int main() {
    //get the first line (this input is only one line)
    std::ifstream file("../inputs/day_2.txt");
    std::string line;
    std::getline(file, line);

    // Split by comma and append to a vector
    std::stringstream ss(line);
    std::string token;
    std::vector<std::string> id_ranges;
    while (std::getline(ss, token, ',')) {
        id_ranges.push_back(token);
    }

    long long total = 0;

    for (const auto& id_range : id_ranges) {
        // split start and end
        size_t dash_pos = id_range.find('-');  // find position of '-'
        std::string start = id_range.substr(0, dash_pos);           // subtring prior to the dash
        std::string end = id_range.substr(dash_pos + 1);            // subtring after the dash
        //we dynamically swap types from int to string and back for iterability & function use
        //we could've also had the function take an int and convert within the function so this is a design choice.
        for(long long i = std::stoll(start); i <= std::stoll(end);i++ ) {
            if(symmetry_checker(std::to_string(i))) {
                total += i;
            }
        }
    }


    std::cout << total << std::endl;
    return 0;
}

File: Python/day_2.py
file = "../inputs/day_2.txt"


def symmetry_checker(s: str) -> bool:
    if len(s) // 2 and s[0:len(s) // 2] == s[len(s) // 2:]:
        return True
    return False


with open(file, "r", encoding="utf-8") as f:  # reading the input
    id_ranges = f.read().split(",")  # this is one line, seperated by commas

    total = 0  # set counter variable for this file run

    # read the file
    for id_range in id_ranges:
        start, end = id_range.split("-")
        start = int(start)
        end = int(end)
        # print(start, end) debug

        for num in range(start, end + 1):
            total += num if symmetry_checker(str(num)) else 0
    print(total)
